Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,213 @@
+* The handle must be closed via `hostfxr_shutdown`.
+* The handle is not thread safe - the consumer should only call functions on it from one thread at a time.
+
+The `hostfxr` will also track active runtime in the process. Due to limitations (and to simplify implementation) this tracking will actually not look at the actual `coreclr` module (or try to communicate with the runtime in any way). Instead `hostfxr` itself will track the host context initialization. The first host context initialization in the process will represent the "loaded runtime". It is only possible to have one "loaded runtime" in the process. Any subsequent host context initialization will just "attach" to the "loaded runtime" instead of creating a new one.
+
+*Note: It is technically possible to initialize host context, and never call any "run" method on it, which would mean the runtime would not get loaded. Trying to initialize a second host context while the first one didn't start the runtime yet will for now be considered illegal. It's a pending issue to solve this problem, as COM activation will probably need to be able to safely initialize/run from multiple threads at a time.*
+
+``` C++
+using hostfxr_handle = void *;
+
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;
+    const char_t * dotnet_root;
+};
+```
+
+The `hostfxr_initialize_parameters` structure stores parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the hosting components. It's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C++
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path  ,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
+
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments.
```

Comment:
>This can be `nullptr` if the app is specified in the command line arguments.

Do we really want this behavior? It reminds me of Win32's [`CreateProcess()`](https://docs.microsoft.com/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa) and the constant confusion between the `lpApplicationName` and `lpCommandLine` parameters. Take a few and read the documentation on those - they are amazing.

Reply:
Well - this is actually describing the current behavior of the `hostfxr` APIs. It comes from the desire to be able to use these APIs for all scenarios. We can limit this I guess.. to make the API cleaner.

Reply:
In a way we should support the case where everything comes from the command line - if I would want to write a host which runs apps (basically a second `dotnet.exe`) that's exactly what would be needed, since the command line can contain host specific options before the app name. For example `dotnet.exe --roll-forward Major app.dll`. So asking the native host code to be able to parse this correct would be tricky.

But maybe it's not a scenario just yet (I'm not entire sure about ASP.NET, it might be required there).

Reply:
I am fine with it, if those are the existing semantics. I just wouldn't want to create a new API with that kind of behavior.

Reply:
I don't think we need the super specific description like `CreateProcess`. In our case it could be as simple as:
* non-null - this application will be executed
* null - the argv/argc specify the command in the format supported by `dotnet.exe` (doc link). Maybe mentioning that we would not want to support SDK commands this way.

We might do this as two APIs, but then those are very similar.